File: lib/api/scim.js
var _ = require('lodash'),
  Promise = require('bluebird'),
  util = require('../util'),
  log = util.logger();

function Scim(hootsuite, connection) {
  this._hootsuite = hootsuite;
  this._connection = connection;
}

Scim.prototype = {
  createUser: function (msg) {
    var path = util.createScimPath('Users');
    var data = {
      schemas: msg.schemas || undefined,
      userName: msg.userName || undefined,
      name: msg.name || undefined,
      emails: msg.emails || undefined,
      displayName: msg.displayName || undefined,
      timezone: msg.timezone || undefined,
      preferredLanguage: msg.preferredLanguage || undefined,
      groups: msg.groups || undefined,
      active: msg.active || true,
      'urn:ietf:params:scim:schemas:extension:Hootsuite:2.0:User': msg['urn:ietf:params:scim:schemas:extension:Hootsuite:2.0:User'] || undefined
    };
    return this._connection.postJson(path, data);
  },
  findUsers: function (filter, count, startIndex) {
    var path = util.createScimPath('Users');
    var options = {
      query: {
        filter: filter || undefined,
        count: count || undefined,
        startIndex: startIndex || undefined,
      },
      _method: 'GET'
    };
    return this._connection.get(path, options);
  },
  findUserById: function (memberId) {
    var path = util.createScimPath('Users', memberId);
    return this._connection.get(path, { _method: 'GET' });
  },
  replaceUserById: function (memberId, msg) {
    var path = util.createScimPath('Users', memberId);
    var data = {
      schemas: msg.schemas || undefined,
      userName: msg.userName || undefined,
      name: msg.name || undefined,
      emails: msg.emails || undefined,
      displayName: msg.displayName || undefined,
      timezone: msg.timezone || undefined,
      preferredLanguage: msg.preferredLanguage || undefined,
      groups: msg.groups || undefined,
      active: msg.active || true,
      'urn:ietf:params:scim:schemas:extension:Hootsuite:2.0:User': msg['urn:ietf:params:scim:schemas:extension:Hootsuite:2.0:User'] || undefined
    };
    return this._connection.putJson(path, data);
  },
  modifyUserById: function (memberId, msg) {
    var path = util.createScimPath('Users', memberId);
    var options = {
      headers: { 'content-type': 'application/json' },
      data: JSON.stringify({
        schemas: msg.schemas || undefined,
        Operations: msg.Operations || undefined,
      }),
      _method: 'PATCH'
    };
    return this._connection.patch(path, options);
  },
  getResourceTypes: function () {
    var path = util.createScimPath('ResourceTypes');
    return this._connection.get(path, { _method: 'GET' });
  },
}

module.exports = Scim;

File: src.net/Hootsuite.Rest/Api/Scim.cs
using Newtonsoft.Json.Linq;
using System.Threading.Tasks;

namespace Hootsuite.Rest.Api
{
    public class Scim
    {
        Hootsuite _hootsuite;
        Connection _connection;

        public Scim(Hootsuite hootsuite, Connection connection)
        {
            _hootsuite = hootsuite;
            _connection = connection;
        }

        public Task<JObject> createUser(dynamic msg)
        {
            var path = util.createScimPath("Users");
            var data = new
            {
                schemas = dyn.getProp(msg, "schemas", (string)null),
                userName = dyn.getProp(msg, "userName", (string)null),
                name = dyn.getProp(msg, "name", (string)null),
                emails = dyn.getProp(msg, "emails", (string)null),
                displayName = dyn.getProp(msg, "displayName", (string)null),
                timezone = dyn.getProp(msg, "timezone", (string)null),
                preferredLanguage = dyn.getProp(msg, "preferredLanguage", (string)null),
                groups = dyn.getProp(msg, "groups", (string)null),
                active = dyn.getProp(msg, "active", true),
                //"urn:ietf:params:scim:schemas:extension:Hootsuite:2.0:User" = msg['urn:ietf:params:scim:schemas:extension:Hootsuite:2.0:User'] ?? null,
            };
            return _connection.postJson(path, data);
        }

        public Task<JObject> findUsers(string filter = null, int count = 0, int startIndex = 0)
        {
            var path = util.createScimPath("Users");
            var options = new
            {
                query = new
                {
                    filter,
                    count,
                    startIndex
                },
            };
            return _connection.get(path, options);
        }

        public Task<JObject> findUserById(string memberId)
        {
            var path = util.createScimPath("Users", memberId);
            return _connection.get(path);
        }

        public Task<JObject> replaceUserById(string memberId, dynamic msg)
        {
            var path = util.createScimPath("Users", memberId);
            var data = new
            {
                schemas = dyn.getProp(msg, "schemas", (string)null),
                userName = dyn.getProp(msg, "userName", (string)null),
                name = dyn.getProp(msg, "name", (string)null),
                emails = dyn.getProp(msg, "emails", (string)null),
                displayName = dyn.getProp(msg, "displayName", (string)null),
                timezone = dyn.getProp(msg, "timezone", (string)null),
                preferredLanguage = dyn.getProp(msg, "preferredLanguage", (string)null),
                groups = dyn.getProp(msg, "groups", (string)null),
                active = dyn.getProp(msg, "active", true),
                //'urn:ietf:params:scim:schemas:extension:Hootsuite:2.0:User': msg['urn:ietf:params:scim:schemas:extension:Hootsuite:2.0:User'] || undefined
            };
            return _connection.putJson(path, data);
        }

        public Task<JObject> modifyUserById(string memberId, dynamic msg)
        {
            var path = util.createScimPath("Users", memberId);
            var options = new
            {
                //headers = new { "content-type" = "application/json" },
                data = new
                {
                    schemas = dyn.getProp(msg, "schemas", (string)null),
                    Operations = dyn.getProp(msg, "Operations", (string)null),
                },
            };
            return _connection.patch(path, options);
        }

        public Task<JObject> getResourceTypes()
        {
            var path = util.createScimPath("ResourceTypes");
            return _connection.get(path);
        }
    }
}
